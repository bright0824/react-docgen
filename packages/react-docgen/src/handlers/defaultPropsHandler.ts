import getPropertyName from '../utils/getPropertyName.js';
import getMemberValuePath from '../utils/getMemberValuePath.js';
import printValue from '../utils/printValue.js';
import resolveToValue from '../utils/resolveToValue.js';
import resolveFunctionDefinitionToReturnValue from '../utils/resolveFunctionDefinitionToReturnValue.js';
import isReactComponentClass from '../utils/isReactComponentClass.js';
import isReactForwardRefCall from '../utils/isReactForwardRefCall.js';
import type Documentation from '../Documentation.js';
import type { DefaultValueDescriptor } from '../Documentation.js';
import type { NodePath } from '@babel/traverse';
import type {
  CallExpression,
  ObjectMethod,
  ObjectProperty,
  RestElement,
  SpreadElement,
} from '@babel/types';
import type {
  ComponentNode,
  StatelessComponentNode,
} from '../resolver/index.js';
import type { Handler } from './index.js';

function getDefaultValue(path: NodePath): DefaultValueDescriptor | null {
  let defaultValue: string | undefined;
  let resolvedPath = path;
  let valuePath = path;

  if (path.isBooleanLiteral()) {
    defaultValue = `${path.node.value}`;
  } else if (path.isNullLiteral()) {
    defaultValue = 'null';
  } else if (path.isLiteral()) {
    defaultValue = path.node.extra?.raw as string;
  } else {
    if (path.isAssignmentPattern()) {
      resolvedPath = resolveToValue(path.get('right'));
    } else {
      resolvedPath = resolveToValue(path);
    }
    if (resolvedPath.isImportDeclaration() && path.isIdentifier()) {
      defaultValue = path.node.name;
    } else {
      valuePath = resolvedPath;
      defaultValue = printValue(resolvedPath);
    }
  }
  if (typeof defaultValue !== 'undefined') {
    return {
      value: defaultValue,
      computed:
        valuePath.isCallExpression() ||
        valuePath.isMemberExpression() ||
        valuePath.isIdentifier(),
    };
  }

  return null;
}

function getStatelessPropsPath(
  componentDefinition: NodePath<CallExpression | StatelessComponentNode>,
): NodePath | undefined {
  let value: NodePath = componentDefinition;

  if (isReactForwardRefCall(componentDefinition)) {
    value = resolveToValue(componentDefinition.get('arguments')[0]!);
  }

  if (!value.isFunction()) {
    return;
  }

  return value.get('params')[0];
}

function getDefaultPropsPath(
  componentDefinition: NodePath<ComponentNode>,
): NodePath | null {
  let defaultPropsPath: NodePath | null = getMemberValuePath(
    componentDefinition,
    'defaultProps',
  );

  if (!defaultPropsPath) {
    return null;
  }

  defaultPropsPath = resolveToValue(defaultPropsPath);
  if (!defaultPropsPath) {
    return null;
  }

  if (
    defaultPropsPath.isFunctionExpression() ||
    defaultPropsPath.isFunctionDeclaration() ||
    defaultPropsPath.isClassMethod() ||
    defaultPropsPath.isObjectMethod()
  ) {
    // Find the value that is returned from the function and process it if it is
    // an object literal.
    const returnValue =
      resolveFunctionDefinitionToReturnValue(defaultPropsPath);

    if (returnValue && returnValue.isObjectExpression()) {
      defaultPropsPath = returnValue;
    }
  }

  return defaultPropsPath;
}

function getDefaultValuesFromProps(
  properties: Array<
    NodePath<ObjectMethod | ObjectProperty | RestElement | SpreadElement>
  >,
  documentation: Documentation,
  isStateless: boolean,
): void {
  properties.forEach((propertyPath) => {
    if (propertyPath.isObjectProperty()) {
      const propName = getPropertyName(propertyPath);

      if (!propName) return;

      let valuePath = propertyPath.get('value');

      if (isStateless) {
        if (valuePath.isAssignmentPattern()) {
          valuePath = valuePath.get('right');
        } else {
          // Don't evaluate property if component is functional and the node is not an AssignmentPattern
          return;
        }
      }

      // Initialize the prop descriptor here after the early return from above
      const propDescriptor = documentation.getPropDescriptor(propName);
      const defaultValue = getDefaultValue(valuePath);

      if (defaultValue) {
        propDescriptor.defaultValue = defaultValue;
      }
    } else if (propertyPath.isSpreadElement()) {
      const resolvedValuePath = resolveToValue(propertyPath.get('argument'));

      if (resolvedValuePath.isObjectExpression()) {
        getDefaultValuesFromProps(
          resolvedValuePath.get('properties'),
          documentation,
          isStateless,
        );
      }
    }
  });
}

const defaultPropsHandler: Handler = function (
  documentation: Documentation,
  componentDefinition: NodePath<ComponentNode>,
): void {
  let statelessProps: NodePath | undefined;
  const defaultPropsPath = getDefaultPropsPath(componentDefinition);

  /**
   * function, lazy, memo, forwardRef etc components can resolve default props as well
   */
  if (!isReactComponentClass(componentDefinition)) {
    statelessProps = getStatelessPropsPath(
      componentDefinition as NodePath<CallExpression | StatelessComponentNode>,
    );
  }

  // Do both statelessProps and defaultProps if both are available so defaultProps can override
  if (statelessProps && statelessProps.isObjectPattern()) {
    getDefaultValuesFromProps(
      statelessProps.get('properties'),
      documentation,
      true,
    );
  }
  if (defaultPropsPath && defaultPropsPath.isObjectExpression()) {
    getDefaultValuesFromProps(
      defaultPropsPath.get('properties'),
      documentation,
      false,
    );
  }
};

export default defaultPropsHandler;
