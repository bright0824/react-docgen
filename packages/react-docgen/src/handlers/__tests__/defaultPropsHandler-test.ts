import type { NodePath } from '@babel/traverse';
import type {
  ArrowFunctionExpression,
  CallExpression,
  ClassDeclaration,
  ClassExpression,
  ObjectExpression,
  VariableDeclaration,
} from '@babel/types';
import { parse, makeMockImporter } from '../../../tests/utils';
import Documentation from '../../Documentation';
import type DocumentationMock from '../../__mocks__/Documentation';
import defaultPropsHandler from '../defaultPropsHandler.js';
import { beforeEach, describe, expect, test, vi } from 'vitest';

vi.mock('../../Documentation.js');

describe('defaultPropsHandler', () => {
  let documentation: Documentation & DocumentationMock;

  beforeEach(() => {
    documentation = new Documentation() as Documentation & DocumentationMock;
  });

  const mockImporter = makeMockImporter({
    getDefaultProps: stmtLast =>
      stmtLast(`
      import baz from 'baz';
      export default function() {
        return {
          foo: "bar",
          bar: 42,
          baz: baz,
          abc: {xyz: abc.def, 123: 42}
        };
      }
    `).get('declaration'),

    baz: stmtLast =>
      stmtLast(`
      export default ["foo", "bar"];
    `).get('declaration'),

    other: stmtLast =>
      stmtLast(`
      export default { bar: "foo" };
    `).get('declaration'),

    defaultProps: stmtLast =>
      stmtLast(`
      export default {
        foo: "bar",
        bar: 42,
        baz: ["foo", "bar"],
        abc: {xyz: abc.def, 123: 42}
      };
    `).get('declaration'),
  });

  describe('ObjectExpression', () => {
    test('should find prop default values that are literals', () => {
      const src = `
        {
          getDefaultProps: function() {
            return {
              foo: "bar",
              bar: 42,
              falseliteral: false,
              trueliteral: true,
              nullliteral: null,
              regex: /./,
              bigint: 1n,
              baz: ["foo", "bar"],
              abc: {xyz: abc.def, 123: 42}
            };
          }
        }
      `;

      defaultPropsHandler(
        documentation,
        parse.expression<ObjectExpression>(src),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('can resolve object methods', () => {
      const src = `
        {
          getDefaultProps() {
            return {
              foo: "bar",
            };
          }
        }
      `;

      defaultPropsHandler(
        documentation,
        parse.expression<ObjectExpression>(src),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('can resolve declared functions', () => {
      const src = `
        function getDefaultProps() {
          return {
            foo: "bar",
            bar: 42,
            baz: ["foo", "bar"],
            abc: {xyz: abc.def, 123: 42}
          };
        }
        ({
          getDefaultProps: getDefaultProps
        })
      `;

      defaultPropsHandler(documentation, parse.expressionLast(src));
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('should find prop default values that are literals from imported functions', () => {
      const src = `
        import getDefaultProps from 'getDefaultProps';

        ({
          getDefaultProps: getDefaultProps
        })
      `;

      defaultPropsHandler(
        documentation,
        parse.expressionLast(src, mockImporter),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('handles computed properties', () => {
      const src = `
        {
          getDefaultProps: function() {
            return {
              foo: "bar",
              [bar]: 42,
            };
          }
        }
      `;

      defaultPropsHandler(
        documentation,
        parse.expression<ObjectExpression>(src),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('handles imported values assigned to computed properties', () => {
      const src = `
        import baz from 'baz';
        ({
          getDefaultProps: function() {
            return {
              foo: "bar",
              [bar]: baz,
            };
          }
        })
      `;

      defaultPropsHandler(
        documentation,
        parse.expressionLast(src, mockImporter),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('ignores complex computed properties', () => {
      const src = `
        ({
          getDefaultProps: function() {
            return {
              foo: "bar",
              [() => {}]: 42,
            };
          }
        })
      `;

      defaultPropsHandler(documentation, parse.expressionLast(src));
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('ignores imported values assigned to complex computed properties', () => {
      const src = `
        import baz from 'baz';
        ({
          getDefaultProps: function() {
            return {
              foo: "bar",
              [() => {}]: baz,
            };
          }
        })
      `;

      defaultPropsHandler(
        documentation,
        parse.expressionLast(src, mockImporter),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('resolves local spreads', () => {
      const src = `
        const other = { bar: "foo" };

        ({
          getDefaultProps: function() {
            return {
              foo: "bar",
              ...other,
            };
          }
        })
      `;

      defaultPropsHandler(documentation, parse.expressionLast(src));
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('resolves imported spreads', () => {
      const src = `
        import other from 'other';
        ({
          getDefaultProps: function() {
            return {
              foo: "bar",
              ...other,
            };
          }
        })
      `;

      defaultPropsHandler(
        documentation,
        parse.expressionLast(src, mockImporter),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });
  });

  describe('ClassDeclaration with static defaultProps', () => {
    test('should find prop default values that are literals', () => {
      const src = `
        class Foo {
          static defaultProps = {
            foo: "bar",
            bar: 42,
            baz: ["foo", "bar"],
            abc: {xyz: abc.def, 123: 42}
          };
        }
      `;

      defaultPropsHandler(
        documentation,
        parse.statement<ClassDeclaration>(src),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('should find prop default values that are literals', () => {
      const src = `
        class Foo {
          static get defaultProps() {
            return {
              foo: "bar",
            };
          }
        }
      `;

      defaultPropsHandler(
        documentation,
        parse.statement<ClassDeclaration>(src),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('resolves imported values assigned as default props', () => {
      const src = `
        import defaultProps from 'defaultProps';
        class Foo {
          static defaultProps = defaultProps;
        }
      `;

      defaultPropsHandler(
        documentation,
        parse.statementLast(src, mockImporter),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('should resolve local spreads', () => {
      const src = `
        const other = { bar: "foo" };

        class Foo {
          static defaultProps = {
            foo: "bar",
            ...other
          };
        }
      `;

      defaultPropsHandler(documentation, parse.statementLast(src));
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('resolves imported spreads', () => {
      const src = `
        import other from 'other';
        class Foo {
          static defaultProps = {
            foo: "bar",
            ...other
          };
        }
      `;

      defaultPropsHandler(
        documentation,
        parse.statementLast(src, mockImporter),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('should find prop default values that are imported variables', () => {
      const src = `
        import ImportedComponent from './ImportedComponent.js';

        class Foo {
          static defaultProps = {
            foo: ImportedComponent,
          };
        }
      `;

      defaultPropsHandler(documentation, parse.statementLast(src));
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('can resolve default props that are imported given a custom importer', () => {
      const src = `
        import baz from 'baz';

        class Foo {
          static defaultProps = {
            baz: baz,
          };
        }
      `;

      defaultPropsHandler(
        documentation,
        parse.statementLast(src, mockImporter),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });
  });

  describe('ClassExpression with static defaultProps', () => {
    test('should find prop default values that are literals', () => {
      const src = `
        var Bar = class {
          static defaultProps = {
            foo: "bar",
            bar: 42,
            baz: ["foo", "bar"],
            abc: {xyz: abc.def, 123: 42}
          };
      }`;

      defaultPropsHandler(
        documentation,
        parse
          .statement(src)
          .get('declarations.0.init') as NodePath<ClassExpression>,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('resolves imported values assigned as default props', () => {
      const src = `
        import defaultProps from 'defaultProps';
        var Bar = class {
          static defaultProps = defaultProps;
        }
      `;

      defaultPropsHandler(
        documentation,
        parse
          .statementLast(src, mockImporter)
          .get('declarations.0.init') as NodePath<ClassExpression>,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });
  });

  test('should only consider Property nodes, not e.g. spread properties', () => {
    const src = `
      {
        getDefaultProps: function() {
          return {
            ...Foo.bar,
            bar: 42,
          };
        }
      }
    `;
    const definition = parse.expression<ObjectExpression>(src);

    expect(() => defaultPropsHandler(documentation, definition)).not.toThrow();
    expect(documentation.descriptors).toMatchSnapshot();
  });

  test('can have an importer that resolves spread properties', () => {
    const src = `
      import Props from 'defaultProps';
      ({
        getDefaultProps: function() {
          return {
            ...Props.abc,
            bar: 42,
          };
        }
      })
    `;
    const definition = parse.expressionLast<ObjectExpression>(
      src,
      mockImporter,
    );

    expect(() => defaultPropsHandler(documentation, definition)).not.toThrow();
    expect(documentation.descriptors).toMatchSnapshot();
  });

  describe('Functional components with default params', () => {
    test('should find default props that are literals', () => {
      const src = `
        ({
          foo = "bar",
          bar = 42,
          baz = ["foo", "bar"],
          abc = {xyz: abc.def, 123: 42}
        }) => <div />
      `;

      defaultPropsHandler(documentation, parse.expressionLast(src));
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('can use imported values as default props', () => {
      const src = `
        import baz from 'baz';
        ({
          bar = baz,
        }) => <div />
      `;

      defaultPropsHandler(
        documentation,
        parse.expressionLast(src, mockImporter),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('should override with defaultProps if available', () => {
      const src = `
        var Foo = ({
          foo = "bar",
          bar = 42,
          baz = ["foo", "bar"],
          abc = 'test'
        }) => <div />
        Foo.defaultProps = { abc: {xyz: abc.def, 123: 42} };
      `;

      defaultPropsHandler(
        documentation,
        parse
          .statement(src)
          .get('declarations.0.init') as NodePath<ArrowFunctionExpression>,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('overrides with imported defaultProps', () => {
      const src = `
        import other from 'other';
        var Foo = ({
          bar = 42,
        }) => <div />
        Foo.defaultProps = other;
      `;

      defaultPropsHandler(
        documentation,
        parse
          .statement(src, mockImporter, 1)
          .get('declarations.0.init') as NodePath<ArrowFunctionExpression>,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('resolves local spreads', () => {
      const src = `
        const other = { bar: "foo" };
        var Foo = (props) => <div />
        Foo.defaultProps = { foo: "bar", ...other };
      `;

      defaultPropsHandler(
        documentation,
        parse
          .statement(src, 1)
          .get('declarations.0.init') as NodePath<ArrowFunctionExpression>,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('resolves imported spreads', () => {
      const src = `
        import other from 'other';
        var Foo = (props) => <div />
        Foo.defaultProps = { foo: "bar", ...other };
      `;

      defaultPropsHandler(
        documentation,
        parse
          .statement(src, mockImporter, 1)
          .get('declarations.0.init') as NodePath<ArrowFunctionExpression>,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('should work with aliases', () => {
      const src = `
        ({
          foo = "bar",
          bar = 42,
          baz = ["foo", "bar"],
          abc: defg = {xyz: abc.def, 123: 42}
        }) => <div />
      `;

      defaultPropsHandler(
        documentation,
        parse.expressionLast<ArrowFunctionExpression>(src),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('allows imported defaults to be aliased', () => {
      const src = `
        import baz from 'baz';
        ({
          foo: bar = baz,
        }) => <div />
      `;

      defaultPropsHandler(
        documentation,
        parse.expressionLast<ArrowFunctionExpression>(src, mockImporter),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('should find prop default values that are imported variables', () => {
      const src = `
        import ImportedComponent from './ImportedComponent.js';

        ({
          foo = ImportedComponent,
        }) => <div />
      `;

      defaultPropsHandler(
        documentation,
        parse.expressionLast<ArrowFunctionExpression>(src),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('should work with no defaults', () => {
      const src = `({ foo }) => <div />`;

      defaultPropsHandler(
        documentation,
        parse.expressionLast<ArrowFunctionExpression>(src),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });
  });

  describe('forwardRef', () => {
    test('resolves default props in the parameters', () => {
      const src = `
        import React from 'react';
        React.forwardRef(({ foo = 'bar' }, ref) => <div ref={ref}>{foo}</div>);
      `;

      defaultPropsHandler(
        documentation,
        parse.expressionLast<CallExpression>(src),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('resolves imported default props in the parameters', () => {
      const src = `
        import baz from 'baz';
        import React from 'react';
        React.forwardRef(({ bar = baz }, ref) => <div ref={ref}>{bar}</div>);
      `;

      defaultPropsHandler(
        documentation,
        parse.expressionLast<CallExpression>(src, mockImporter),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('resolves defaultProps', () => {
      const src = `
        import React from 'react';
        const Component = React.forwardRef(({ foo }, ref) => <div ref={ref}>{foo}</div>);
        Component.defaultProps = { foo: 'baz' };
      `;

      defaultPropsHandler(
        documentation,
        parse
          .statement<VariableDeclaration>(src, 1)
          .get('declarations.0.init') as NodePath<CallExpression>,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('resolves imported defaultProps', () => {
      const src = `
        import other from 'other';
        import React from 'react';
        const Component = React.forwardRef(({ bar }, ref) => <div ref={ref}>{bar}</div>);
        Component.defaultProps = other;
      `;

      defaultPropsHandler(
        documentation,
        parse
          .statement(src, mockImporter, 2)
          .get('declarations.0.init') as NodePath<CallExpression>,
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('resolves when the function is not inline', () => {
      const src = `
        import React from 'react';
        const ComponentImpl = ({ foo = 'bar' }, ref) => <div ref={ref}>{foo}</div>;
        React.forwardRef(ComponentImpl);
      `;

      defaultPropsHandler(
        documentation,
        parse.expressionLast<CallExpression>(src),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });

    test('also resolves imports when the function is not inline', () => {
      const src = `
        import baz from 'baz';
        import React from 'react';
        const ComponentImpl = ({ bar = baz }, ref) => <div ref={ref}>{bar}</div>;
        React.forwardRef(ComponentImpl);
      `;

      defaultPropsHandler(
        documentation,
        parse.expressionLast<CallExpression>(src, mockImporter),
      );
      expect(documentation.descriptors).toMatchSnapshot();
    });
  });
});
