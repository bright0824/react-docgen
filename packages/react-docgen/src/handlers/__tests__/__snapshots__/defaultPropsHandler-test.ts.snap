// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaultPropsHandler ClassDeclaration with static defaultProps can resolve default props that are imported given a custom importer 1`] = `
{
  "baz": {
    "defaultValue": {
      "computed": false,
      "value": "["foo", "bar"]",
    },
  },
}
`;

exports[`defaultPropsHandler ClassDeclaration with static defaultProps resolves imported spreads 1`] = `
{
  "bar": {
    "defaultValue": {
      "computed": false,
      "value": ""foo"",
    },
  },
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": ""bar"",
    },
  },
}
`;

exports[`defaultPropsHandler ClassDeclaration with static defaultProps resolves imported values assigned as default props 1`] = `
{
  "abc": {
    "defaultValue": {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": {
    "defaultValue": {
      "computed": false,
      "value": "42",
    },
  },
  "baz": {
    "defaultValue": {
      "computed": false,
      "value": "["foo", "bar"]",
    },
  },
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": ""bar"",
    },
  },
}
`;

exports[`defaultPropsHandler ClassDeclaration with static defaultProps should find prop default values that are imported variables 1`] = `
{
  "foo": {
    "defaultValue": {
      "computed": true,
      "value": "ImportedComponent",
    },
  },
}
`;

exports[`defaultPropsHandler ClassDeclaration with static defaultProps should find prop default values that are literals 1`] = `
{
  "abc": {
    "defaultValue": {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": {
    "defaultValue": {
      "computed": false,
      "value": "42",
    },
  },
  "baz": {
    "defaultValue": {
      "computed": false,
      "value": "["foo", "bar"]",
    },
  },
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": ""bar"",
    },
  },
}
`;

exports[`defaultPropsHandler ClassDeclaration with static defaultProps should find prop default values that are literals 2`] = `
{
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": ""bar"",
    },
  },
}
`;

exports[`defaultPropsHandler ClassDeclaration with static defaultProps should resolve local spreads 1`] = `
{
  "bar": {
    "defaultValue": {
      "computed": false,
      "value": ""foo"",
    },
  },
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": ""bar"",
    },
  },
}
`;

exports[`defaultPropsHandler ClassExpression with static defaultProps resolves imported values assigned as default props 1`] = `
{
  "abc": {
    "defaultValue": {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": {
    "defaultValue": {
      "computed": false,
      "value": "42",
    },
  },
  "baz": {
    "defaultValue": {
      "computed": false,
      "value": "["foo", "bar"]",
    },
  },
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": ""bar"",
    },
  },
}
`;

exports[`defaultPropsHandler ClassExpression with static defaultProps should find prop default values that are literals 1`] = `
{
  "abc": {
    "defaultValue": {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": {
    "defaultValue": {
      "computed": false,
      "value": "42",
    },
  },
  "baz": {
    "defaultValue": {
      "computed": false,
      "value": "["foo", "bar"]",
    },
  },
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": ""bar"",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params allows imported defaults to be aliased 1`] = `
{
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": "["foo", "bar"]",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params can use imported values as default props 1`] = `
{
  "bar": {
    "defaultValue": {
      "computed": false,
      "value": "["foo", "bar"]",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params overrides with imported defaultProps 1`] = `
{
  "bar": {
    "defaultValue": {
      "computed": false,
      "value": ""foo"",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params resolves imported spreads 1`] = `
{
  "bar": {
    "defaultValue": {
      "computed": false,
      "value": ""foo"",
    },
  },
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": ""bar"",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params resolves local spreads 1`] = `
{
  "bar": {
    "defaultValue": {
      "computed": false,
      "value": ""foo"",
    },
  },
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": ""bar"",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params should find default props that are literals 1`] = `
{
  "abc": {
    "defaultValue": {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": {
    "defaultValue": {
      "computed": false,
      "value": "42",
    },
  },
  "baz": {
    "defaultValue": {
      "computed": false,
      "value": "["foo", "bar"]",
    },
  },
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": ""bar"",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params should find prop default values that are imported variables 1`] = `
{
  "foo": {
    "defaultValue": {
      "computed": true,
      "value": "ImportedComponent",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params should override with defaultProps if available 1`] = `
{
  "abc": {
    "defaultValue": {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": {
    "defaultValue": {
      "computed": false,
      "value": "42",
    },
  },
  "baz": {
    "defaultValue": {
      "computed": false,
      "value": "["foo", "bar"]",
    },
  },
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": ""bar"",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params should work with aliases 1`] = `
{
  "abc": {
    "defaultValue": {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": {
    "defaultValue": {
      "computed": false,
      "value": "42",
    },
  },
  "baz": {
    "defaultValue": {
      "computed": false,
      "value": "["foo", "bar"]",
    },
  },
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": ""bar"",
    },
  },
}
`;

exports[`defaultPropsHandler Functional components with default params should work with no defaults 1`] = `{}`;

exports[`defaultPropsHandler ObjectExpression can resolve declared functions 1`] = `
{
  "abc": {
    "defaultValue": {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": {
    "defaultValue": {
      "computed": false,
      "value": "42",
    },
  },
  "baz": {
    "defaultValue": {
      "computed": false,
      "value": "["foo", "bar"]",
    },
  },
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": ""bar"",
    },
  },
}
`;

exports[`defaultPropsHandler ObjectExpression can resolve object methods 1`] = `
{
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": ""bar"",
    },
  },
}
`;

exports[`defaultPropsHandler ObjectExpression handles computed properties 1`] = `
{
  "@computed#bar": {
    "defaultValue": {
      "computed": false,
      "value": "42",
    },
  },
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": ""bar"",
    },
  },
}
`;

exports[`defaultPropsHandler ObjectExpression handles imported values assigned to computed properties 1`] = `
{
  "@computed#bar": {
    "defaultValue": {
      "computed": false,
      "value": "["foo", "bar"]",
    },
  },
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": ""bar"",
    },
  },
}
`;

exports[`defaultPropsHandler ObjectExpression ignores complex computed properties 1`] = `
{
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": ""bar"",
    },
  },
}
`;

exports[`defaultPropsHandler ObjectExpression ignores imported values assigned to complex computed properties 1`] = `
{
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": ""bar"",
    },
  },
}
`;

exports[`defaultPropsHandler ObjectExpression resolves imported spreads 1`] = `
{
  "bar": {
    "defaultValue": {
      "computed": false,
      "value": ""foo"",
    },
  },
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": ""bar"",
    },
  },
}
`;

exports[`defaultPropsHandler ObjectExpression resolves local spreads 1`] = `
{
  "bar": {
    "defaultValue": {
      "computed": false,
      "value": ""foo"",
    },
  },
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": ""bar"",
    },
  },
}
`;

exports[`defaultPropsHandler ObjectExpression should find prop default values that are literals 1`] = `
{
  "abc": {
    "defaultValue": {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": {
    "defaultValue": {
      "computed": false,
      "value": "42",
    },
  },
  "baz": {
    "defaultValue": {
      "computed": false,
      "value": "["foo", "bar"]",
    },
  },
  "bigint": {
    "defaultValue": {
      "computed": false,
      "value": "1n",
    },
  },
  "falseliteral": {
    "defaultValue": {
      "computed": false,
      "value": "false",
    },
  },
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": ""bar"",
    },
  },
  "nullliteral": {
    "defaultValue": {
      "computed": false,
      "value": "null",
    },
  },
  "regex": {
    "defaultValue": {
      "computed": false,
      "value": "/./",
    },
  },
  "trueliteral": {
    "defaultValue": {
      "computed": false,
      "value": "true",
    },
  },
}
`;

exports[`defaultPropsHandler ObjectExpression should find prop default values that are literals from imported functions 1`] = `
{
  "abc": {
    "defaultValue": {
      "computed": false,
      "value": "{xyz: abc.def, 123: 42}",
    },
  },
  "bar": {
    "defaultValue": {
      "computed": false,
      "value": "42",
    },
  },
  "baz": {
    "defaultValue": {
      "computed": false,
      "value": "["foo", "bar"]",
    },
  },
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": ""bar"",
    },
  },
}
`;

exports[`defaultPropsHandler can have an importer that resolves spread properties 1`] = `
{
  "123": {
    "defaultValue": {
      "computed": false,
      "value": "42",
    },
  },
  "bar": {
    "defaultValue": {
      "computed": false,
      "value": "42",
    },
  },
  "xyz": {
    "defaultValue": {
      "computed": true,
      "value": "abc.def",
    },
  },
}
`;

exports[`defaultPropsHandler forwardRef also resolves imports when the function is not inline 1`] = `
{
  "bar": {
    "defaultValue": {
      "computed": false,
      "value": "["foo", "bar"]",
    },
  },
}
`;

exports[`defaultPropsHandler forwardRef resolves default props in the parameters 1`] = `
{
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": "'bar'",
    },
  },
}
`;

exports[`defaultPropsHandler forwardRef resolves defaultProps 1`] = `
{
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": "'baz'",
    },
  },
}
`;

exports[`defaultPropsHandler forwardRef resolves imported default props in the parameters 1`] = `
{
  "bar": {
    "defaultValue": {
      "computed": false,
      "value": "["foo", "bar"]",
    },
  },
}
`;

exports[`defaultPropsHandler forwardRef resolves imported defaultProps 1`] = `
{
  "bar": {
    "defaultValue": {
      "computed": false,
      "value": ""foo"",
    },
  },
}
`;

exports[`defaultPropsHandler forwardRef resolves when the function is not inline 1`] = `
{
  "foo": {
    "defaultValue": {
      "computed": false,
      "value": "'bar'",
    },
  },
}
`;

exports[`defaultPropsHandler should only consider Property nodes, not e.g. spread properties 1`] = `
{
  "bar": {
    "defaultValue": {
      "computed": false,
      "value": "42",
    },
  },
}
`;
